lib/fortnox/api/repositories/base.rb

```diff
@@ -1,22 +1,70 @@
-require "fortnox/api/base"
+require 'httparty'
+require 'fortnox/api/request_handling'
 require "fortnox/api/repositories/base/loaders"
 require "fortnox/api/repositories/base/savers"
 
 module Fortnox
   module API
     module Repository
-      class Base < Fortnox::API::Base
-
+      class Base
+        include HTTParty
+        include Fortnox::API::RequestHandling
         include Loaders
         include Savers
 
+        HTTParty::Parser::SupportedFormats[ "text/html" ] = :json
+
+        DEFAULT_HEADERS = {
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json',
+        }.freeze
+
+        HTTP_METHODS = [ :get, :put, :post, :delete ].freeze
+
+        attr_accessor :headers
         attr_reader :mapper, :keys_filtered_on_save
 
-        def initialize( keys_filtered_on_save: [ :url ] )
-          super()
+        def self.set_headers( headers = {} )
+          self.headers.merge!( headers )
+        end
+
+        HTTP_METHODS.each do |method|
+          define_method method do |*args|
+            self.headers['Access-Token'] = next_access_token
+            execute do |remote|
+              remote.send( method, *args )
+            end
+          end
+        end
+
+        def initialize( model, keys_filtered_on_save: [ :url ], token_store: :default )
+          self.class.base_uri( get_base_url )
+
+          self.headers = DEFAULT_HEADERS.merge({
+            'Client-Secret' => get_client_secret,
+          })
 
           @keys_filtered_on_save = keys_filtered_on_save
-          @mapper = Registry[ Mapper::Base.canonical_name_sym( self.class::MODEL )].new
+          @token_store = token_store
+          @mapper = Registry[ Mapper::Base.canonical_name_sym( model )].new
+        end
+
+        def next_access_token
+          @access_tokens ||= CircularQueue.new( *get_access_tokens )
+          @access_tokens.next
+        end
+
+        def check_access_tokens!( tokens )
+          if tokens == nil or tokens.length.zero?
+            fail MissingConfiguration, "You have not provided any access tokens in token store #{ @token_store.inspect }."
+          end
+        end
+
+        def get_access_tokens
+          raw_tokens = config.access_tokens
+          tokens = raw_tokens.is_a?(Hash) ? get_hash_access_tokens(raw_tokens) : raw_tokens
```

Comment:
It would be cleaner if the config took care of this for us, see the comment above for a single access token.

We could define three setters for the config: `access_token`, `access_tokens` and `token_store` where the first takes a string, the second takes an array and the third a hash. The first two would then in actuality insert the config value into the `token_store` config value by appropriately wrapping the value as `{ default: [ value ]}` and `{ default: value }` respectively.

That way we only ever need to handle one case in our code.

Reply:
That would be nice, but where to put this code? This cannot be done in `Fortnox::API` since it's a module. Wouldn't we need to do this logic in `Fortnox::API::Repository::Base#initialize`?

Reply:
I would like that to be supported by the configuration module, but I'm not sure it is :) You will have to play with the processor version of declaring a setting:
```
setting :token_store, {}

setting(:access_token) do |value|
  # some hack to set @_config['token_store'] to { default: [ value ]}
end

setting(:access_tokens) do |value|
  # some hack to set @_config['token_store'] to { default: value }
end
`
``
But as I said, I'm not sure it's possible.

Reply:
I did try something like this, but that did not work.
```ruby
setting :token_store, { default: self.config.access_token }
```
I've also looked at the [gem documentation](http://dry-rb.org/gems/dry-configurable/) without finding anything about it. I'll continue to play around and look for a hack :)

Reply:
I think I've found a solution for this now :) Implementation on the way.